File: src/main/kotlin/domain/account/data/AccountInfoRepo.java
package domain.account.data;

import org.elegantobjects.jpages.LibraryApp.common.util.Result;
import org.elegantobjects.jpages.LibraryApp.common.util.log.ILog;
import org.elegantobjects.jpages.LibraryApp.common.util.uuid2.UUID2;
import org.elegantobjects.jpages.LibraryApp.domain.account.Account;
import org.elegantobjects.jpages.LibraryApp.domain.common.data.repo.Repo;
import org.jetbrains.annotations.NotNull;

/**
 * AccountInfoRepo is a Repo for AccountInfo objects.<br>
 * <br>
 * Holds Account info for all the users in the system (simple CRUD operations)<br>
 * <br>
 * Simulates a database on a server via in-memory HashMap.<br>
 *
 * @author Chris Athanas (realityexpanderdev@gmail.com)
 * @since 0.11
 */

public class AccountInfoRepo extends Repo implements IAccountInfoRepo {

    // simulate a local database on server (UUID2<Account> is the key)
    private final UUID2.HashMap<UUID2<Account>, AccountInfo> database = new UUID2.HashMap<>();

    public
    AccountInfoRepo(@NotNull ILog log) {
        super(log);
    }

    public Result<AccountInfo> fetchAccountInfo(UUID2<Account> id) {
        log.d(this, "id: " + id);

        // Simulate network/database
        if (database.containsKey(id)) {
            return new Result.Success<>(database.get(id));
        }

        return new Result.Failure<>(new Exception("Repo.AccountInfo, account not found, id: " + id));
    }

    @Override
    public Result<AccountInfo> updateAccountInfo(@NotNull AccountInfo accountInfo) {
        log.d(this, "accountInfo.id: " + accountInfo.id());

        // Simulate network/database
        if (database.containsKey(accountInfo.id())) {
            database.put(accountInfo.id(), accountInfo);

            return new Result.Success<>(accountInfo);
        }

        return new Result.Failure<>(new Exception("Repo.AccountInfo, account not found, id: " + accountInfo.id()));
    }

    @Override
    public Result<AccountInfo> upsertAccountInfo(@NotNull AccountInfo accountInfo) {
        log.d(this, "accountInfo.id(): " + accountInfo.id());

        // Simulate network/database
        database.put(accountInfo.id(), accountInfo);

        return new Result.Success<>(accountInfo);
    }
}

File: src/main/kotlin/domain/account/data/AccountInforepo.kt
package com.realityexpander.domain.account.data

import com.realityexpander.common.data.local.JsonFileDatabase
import common.log.ILog
import common.uuid2.UUID2
import domain.account.Account
import domain.account.data.AccountInfo
import domain.account.data.IAccountInfoRepo
import domain.common.data.repo.Repo
import kotlinx.coroutines.runBlocking


/**
 * AccountInfoRepo
 *
 * Simulates a persistent database using a local json file for the AccountInfo.
 *
 * Note: Use Domain-specific language to define the API methods.
 *
 * @author Chris Athanas (realityexpanderdev@gmail.com)
 * @since 0.12
 */

class AccountInfoRepo(
    log: ILog,
    accountInfoRepoDatabaseFilename: String = DEFAULT_LIBRARYINFO_REPO_DATABASE_FILENAME,

    // Use a file database to persist the account info
    private val database: JsonFileDatabase<Account, AccountInfo> = object :
        JsonFileDatabase<Account, AccountInfo>(
            accountInfoRepoDatabaseFilename,
            AccountInfo.serializer()
        ) {
        init {
            runBlocking {
                super.loadFileDatabase()
            }
        }

        override suspend fun updateLookupTables() { /* no-op */ }
    }

) : Repo(log), IAccountInfoRepo {

    override suspend fun fetchAccountInfo(id: UUID2<Account>): Result<AccountInfo> {
        return try {
            Result.success(database.findEntityById(id)
                ?: throw Exception("Account not found, id: $id")
            )
        } catch (e: Exception) {
            return Result.failure(e)
        }
    }

    override suspend fun fetchAllAccountInfo(): Result<List<AccountInfo>> {
        return try {
            Result.success(database.findAllEntities())
        } catch (e: Exception) {
            return Result.failure(e)
        }
    }

    override suspend fun updateAccountInfo(accountInfo: AccountInfo): Result<AccountInfo> {
        return try {
            Result.success(database.updateEntity(accountInfo)
                ?: throw Exception("Account not found, id: ${accountInfo.id()}"))
        } catch (e: Exception) {
            return Result.failure(e)
        }
    }

    override suspend fun upsertAccountInfo(accountInfo: AccountInfo): Result<AccountInfo> {
        return try {
            Result.success(database.upsertEntity(accountInfo)
                ?: throw Exception("Account upsert error, id: ${accountInfo.id()}"))
        } catch (e: Exception) {
            return Result.failure(e)
        }
    }

    override suspend fun deleteAccountInfo(accountInfo: AccountInfo): Result<Unit> {
        return try {
            Result.success(database.deleteEntity(accountInfo))
        } catch (e: Exception) {
            return Result.failure(e)
        }
    }

    override suspend fun deleteDatabase(): Result<Unit> {
        return try {
            Result.success(database.deleteDatabaseFile())
        } catch (e: Exception) {
            return Result.failure(e)
        }
    }

    companion object {
        const val DEFAULT_LIBRARYINFO_REPO_DATABASE_FILENAME = "accountInfoRepoDB.json"
    }
}